File: tests/live-exchange/verify.go
package main

// verify — post-run consistency for the hybrid exchange: conservation is
// checked from the OxiMem balances (never expired) and the durable OxiDB
// ledger is cross-checked against the cumulative trade counter.

import (
	"fmt"
	"math"
	"os"
	"strconv"
)

func verify() int {
	mem, err := DialResp()
	if err != nil {
		panic(err)
	}
	doc, err := Dial()
	if err != nil {
		panic(err)
	}
	ledgerTTL := 120
	if v := os.Getenv("LEDGER_TTL_SECS"); v != "" {
		ledgerTTL, _ = strconv.Atoi(v)
	}
	fails := 0
	check := func(cond bool, msg string) {
		if cond {
			fmt.Println("  OK  " + msg)
		} else {
			fmt.Println("FAIL  " + msg)
			fails++
		}
	}

	fmt.Printf("\n=== exchange verification (hybrid: OxiMem market + OxiDB ledger) ===\n")

	// Cumulative fills settled atomically in OxiMem.
	cntV, _ := mem.Do("GET", "trades:count")
	counter, _ := strconv.Atoi(str(cntV))
	nLedger := doc.Count("trades", map[string]any{})
	qlenV, _ := mem.Do("LLEN", "trades:q")
	qlen, _ := qlenV.(int64)
	fmt.Printf("settled=%d ledger=%d queue_backlog=%d\n", counter, nLedger, qlen)

	check(counter > 0, "trades were executed (traders formed a market)")

	// Ledger completeness: with no TTL every settled fill must be in OxiDB
	// (minus whatever is still queued); with a TTL the ledger is a rolling
	// window, so only require it to be non-empty and bounded by the counter.
	if ledgerTTL == 0 {
		check(nLedger+int(qlen) == counter,
			fmt.Sprintf("ledger complete (%d + %d queued == %d settled)", nLedger, qlen, counter))
	} else {
		check(nLedger > 0 && nLedger <= counter,
			fmt.Sprintf("ledger is a live rolling window (%d of %d, ttl=%ds)", nLedger, counter, ledgerTTL))
	}

	// Distinct uids in the ledger (exactly-once writes; unique index backs it).
	d := doc.Aggregate("trades", []any{
		map[string]any{"$group": map[string]any{"_id": "$uid"}},
		map[string]any{"$count": "n"},
	})
	nDistinct := 0
	if len(d) > 0 {
		nDistinct = int(getF(d[0], "n"))
	}
	check(nDistinct == nLedger, fmt.Sprintf("ledger uids all distinct (%d == %d)", nDistinct, nLedger))

	// Conservation from OxiMem balances (the source of truth).
	totalUSD := 0.0
	negCash := 0
	for u := 0; u < nUsers; u++ {
		v, _ := mem.Do("GET", fmt.Sprintf("usd:user-%d", u))
		f, _ := strconv.ParseFloat(str(v), 64)
		totalUSD += f
		if f < -1e-6 {
			negCash++
		}
	}
	wantUSD := float64(nUsers) * startCash
	check(math.Abs(totalUSD-wantUSD) < 1.0,
		fmt.Sprintf("total USD conserved (%.2f == %.0f) — no money created/destroyed", totalUSD, wantUSD))
	check(negCash == 0, fmt.Sprintf("no negative USD (overdrafts: %d)", negCash))

	symOK, negHold := true, 0
	wantHold := float64(nUsers) * initHolding
	for _, s := range symbols {
		tot := 0.0
		for u := 0; u < nUsers; u++ {
			v, _ := mem.Do("GET", fmt.Sprintf("ast:%s:user-%d", s, u))
			f, _ := strconv.ParseFloat(str(v), 64)
			tot += f
			if f < -1e-6 {
				negHold++
			}
		}
		if math.Abs(tot-wantHold) > 0.01 {
			symOK = false
			fmt.Printf("      %s holdings %.4f != %.0f\n", s, tot, wantHold)
		}
	}
	check(symOK, "each symbol's total holdings conserved (closed system)")
	check(negHold == 0, fmt.Sprintf("no negative holdings / no naked shorts (%d)", negHold))

	// Prices moved from seed → the market was formed by traders.
	moved := 0
	for _, s := range symbols {
		v, _ := mem.Do("GET", "px:"+s)
		f, _ := strconv.ParseFloat(str(v), 64)
		if math.Abs(f-seedPrice[s]) > 1e-9 {
			moved++
		}
	}
	check(moved > 0, fmt.Sprintf("prices set by trading (%d/%d symbols moved from seed)", moved, len(symbols)))

	fmt.Println()
	if fails > 0 {
		fmt.Printf("RESULT: %d CHECK(S) FAILED\n", fails)
		return 1
	}
	fmt.Println("RESULT: ALL CHECKS PASSED — RAM-matched market, durable ledger consistent")
	return 0
}

File: tests/live-exchange/verify.py
#!/usr/bin/env python3
"""Post-run consistency checks over the live-exchange state."""
import sys

sys.path.insert(0, __file__.rsplit("/", 1)[0])
from oxidb_client import OxiDB

db = OxiDB()
fails = []


def check(cond, msg):
    print(("  OK  " if cond else "FAIL  ") + msg)
    if not cond:
        fails.append(msg)


trades = db.find("trades", {})
journal = db.find("journal", {})
receipts = db.find("receipts", {})
accounts = db.find("accounts", {})

print(f"\n=== live-exchange verification ===")
print(f"trades={len(trades)} journal_lines={len(journal)} "
      f"receipts={len(receipts)} accounts={len(accounts)}")

# 1. Idempotency: exactly one receipt per trade uid, uids unique.
trade_uids = [t["uid"] for t in trades]
check(len(trade_uids) == len(set(trade_uids)), "trade uids are unique (no double-fill)")
receipt_uids = set(r["uid"] for r in receipts)
check(set(trade_uids) == receipt_uids,
      "every trade has exactly one idempotency receipt and vice versa")

# 2. Every trade fully present across collections (2 journal lines each).
jl = {}
for j in journal:
    jl[j["uid"]] = jl.get(j["uid"], 0) + 1
partial = [u for u in trade_uids if jl.get(u, 0) != 2]
check(not partial, f"every trade has its 2 journal legs (partial: {len(partial)})")

# 3. Journal cash legs sum to zero net across the system? No — money
#    leaves USD and becomes asset value. Instead: each user's USD balance
#    == start + sum of their USD journal deltas (balance reproducible).
START_CASH = 1_000_000
usd_net = {}
for j in journal:
    if j.get("acct") == "USD":
        usd_net[j["owner"]] = usd_net.get(j["owner"], 0.0) + float(j["delta"])
usd_ok = True
for a in accounts:
    if a["asset"] == "USD":
        expected = START_CASH + usd_net.get(a["owner"], 0.0)
        if abs(float(a["bal"]) - expected) > 1e-6:
            usd_ok = False
            print(f"      {a['owner']} USD {a['bal']} != {expected}")
check(usd_ok, "every user's USD balance is reproducible from the journal")

# 4. Asset positions reproducible from journal too.
asset_net = {}
for j in journal:
    if "delta_asset" in j:
        asset_net[(j["owner"], j["acct"])] = \
            asset_net.get((j["owner"], j["acct"]), 0.0) + float(j["delta_asset"])
pos_ok = True
for a in accounts:
    if a["asset"] != "USD":
        expected = asset_net.get((a["owner"], a["asset"]), 0.0)
        if abs(float(a["bal"]) - expected) > 1e-6:
            pos_ok = False
            print(f"      {a['owner']}/{a['asset']} {a['bal']} != {expected}")
check(pos_ok, "every asset position is reproducible from the journal")

# 5. No negative USD cash (no overdraft — the sufficiency check held).
neg = [a for a in accounts if a["asset"] == "USD" and float(a["bal"]) < -1e-6]
check(not neg, f"no user overdrew USD (negative balances: {len(neg)})")

# 6. Portfolio value conserved: total USD cash + total asset USD-value
#    (from journal usd_value) == users * START_CASH. Since each fill's
#    cash delta and asset usd_value are equal-and-opposite, the sum of
#    all USD cash across users must equal START_CASH*users minus the net
#    USD tied up in assets — which equals the summed asset usd_value.
total_usd = sum(float(a["bal"]) for a in accounts if a["asset"] == "USD")
usd_in_assets = -sum(float(j["delta"]) for j in journal if j.get("acct") == "USD")
n_users = sum(1 for a in accounts if a["asset"] == "USD")
check(abs(total_usd + usd_in_assets - n_users * START_CASH) < 1e-3,
      "portfolio value conserved (cash + USD tied in positions == start)")

print()
if fails:
    print(f"RESULT: {len(fails)} CHECK(S) FAILED")
    sys.exit(1)
print("RESULT: ALL CHECKS PASSED — the ledger is consistent after live trading")
